src/Shared/DiagnosticSourceInstrumentation/PropertyFetcher.cs

```diff
@@ -85,54 +73,109 @@ public bool TryFetch(object obj, out T value, bool skipObjNullCheck = false)
+                }
+
+                // Separated as a local function to be able to target the suppression to just this call.
+                // IL3050 was generated here because of the call to MakeGenericType, which is problematic in AOT if one of the type parameters is a value type;
+                // because the compiler might need to generate code specific to that type.
+                // If the type parameter is a reference type, there will be no problem; because the generated code can be shared among all reference type instantiations.
+#if NET6_0_OR_GREATER
+                [RequiresUnreferencedCode(TrimCompatibilityMessage)]
+                [UnconditionalSuppressMessage("AOT", "IL3050", Justification = "The code guarantees that all the generic parameters are reference types.")]
+#endif
+                static PropertyFetch? DynamicInstantiationHelper(Type declaringType, PropertyInfo propertyInfo)
+                {
+                    return (PropertyFetch?)typeof(PropertyFetch)
+                        .GetMethod(nameof(CreateInstantiated), BindingFlags.NonPublic | BindingFlags.Static)!
+                        .MakeGenericMethod(declaringType) // This is validated in the earlier call chain to be a reference type.
+                        .Invoke(null, new object[] { propertyInfo })!;
+                }
             }
         }
 
-        public virtual bool TryFetch(object obj, out T value)
+        public virtual bool TryFetch(object obj, out T? value)
         {
             value = default;
             return false;
         }
 
-        private sealed class TypedPropertyFetch<TDeclaredObject, TDeclaredProperty> : PropertyFetch
-            where TDeclaredProperty : T
+        // Goal: make PropertyFetcher AOT-compatible.
+        // AOT compiler can't guarantee correctness when call into MakeGenericType or MakeGenericMethod
+        // if one of the generic parameters is a value type (reference types are OK.)
+        // For PropertyFetcher, the decision was made to only support reference type payloads, i.e.:
+        // the object from which to get the property value MUST be a reference type.
+        // Create generics with the declared object type as a generic parameter is OK, but we need the return type
+        // of the property to be a value type (on top of reference types.)
+        // Normally, we would have a helper class like `PropertyFetchInstantiated` that takes 2 generic parameters,
+        // the declared object type, and the type of the property value.
+        // But that would mean calling MakeGenericType, with value type parameters which AOT won't support.
+        //
+        // As a workaround, Generic instantiation was split into:
+        // 1. The property value type comes from the PropertyFetcher generic parameter.
+        //    Compiler supports it even if it is a value type; the type is known statically during compilation
+        //    since PropertyFetcher is used with it.
```

Comment:
This is a bit confusing because "value type" means two different things (I think) in this sentence. Maybe change the first one?

```csharp
        // 1. The object type comes from the PropertyFetcher generic parameter.
        //    Compiler supports it even if it is a value type; the type is known statically during compilation
        //    since PropertyFetcher is used with it.
```

Reply:
Instead of saying "Then, the declared object type **was** passed as a generic parameter" maybe say "Then, the declared object type **is** passed as a generic parameter" (the past tense "was" made it confusing for me)